lib/client/keystore.go

```diff
@@ -687,3 +693,65 @@ func (noLocalKeyStore) SaveCerts(proxy string, cas []auth.TrustedCerts) error {
 }
 func (noLocalKeyStore) GetCerts(proxy string) (*x509.CertPool, error) { return nil, errNoLocalKeyStore }
 func (noLocalKeyStore) GetCertsPEM(proxy string) ([][]byte, error)    { return nil, errNoLocalKeyStore }
+
+type usernameProxyPair struct {
+	username string
+	proxy    string
+}
+
+// MemLocalKeyStore is an in-memory session keystore implementation.
+type MemLocalKeyStore struct {
+	fsLocalNonSessionKeyStore
+	inMem map[usernameProxyPair]*Key
+}
+
+// NewMemLocalKeyStore initializes a MemLocalKeyStore, the key directory here is only used
+// for storing CA certificates and known host fingerprints.
+func NewMemLocalKeyStore(dirPath string) (*MemLocalKeyStore, error) {
+	dirPath, err := initKeysDir(dirPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inMem := make(map[usernameProxyPair]*Key)
+	return &MemLocalKeyStore{fsLocalNonSessionKeyStore: fsLocalNonSessionKeyStore{
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: teleport.ComponentKeyStore,
+		}),
+		KeyDir: dirPath,
+	}, inMem: inMem}, nil
+}
+
+// AddKey writes a key to the underlying key store if MemLocalKeyStore was initialized with saveNewKeys set to true.
+func (s *MemLocalKeyStore) AddKey(proxy string, username string, key *Key) error {
+	s.inMem[usernameProxyPair{username, proxy}] = key
+	return nil
+}
+
+// GetKey returns the session key for the given username and proxy.
+func (s *MemLocalKeyStore) GetKey(proxy, username string, opts ...KeyOption) (*Key, error) {
+	entry, ok := s.inMem[usernameProxyPair{username, proxy}]
```

Comment:
nit: it's pretty easy to accidentally change the semantics by swapping fields with identical types:
```go
entry, ok := s.inMem[usernameProxyPair{username: username, proxy: proxy}]
```
same for `DeleteKey`

Reply:
nit: Only `DeleteKey` is now actually using the struct keys.